lib/firenest/topology/erlang.ex

```diff
@@ -133,112 +126,316 @@ defmodule Firenest.Topology.Erlang do
     end
   end
 
-  defp delete_node_and_notify(%{nodes: nodes, topology: topology, subscribers: subscribers} = state,
-                              node) do
+  defp diff_monitors(state, node, id, remote_names, monitors) when is_list(monitors) do
+    {added, removed} =
+      Enum.reduce(monitors, {[], remote_names}, fn {ref, name}, {added, removed} ->
+        case remote_names do
+          %{^name => ^ref} ->
+            {added, Map.delete(removed, ref)}
+          %{} ->
+            {[name | added], removed}
+        end
+      end)
+
+    %{local_names: local_names} = state
+
+    for {name, _ref} <- removed do
+      local_monitor_down(local_names, node, id, name)
+    end
+
+    for name <- added do
+      local_monitor_up(local_names, node, id, name)
+    end
+
+    :ok
+  end
+
+  defp add_node_and_notify(state, node, id, pid, remote_names, monitors) do
+    %{
+      topology: topology,
+      nodes: nodes,
+      local_names: local_names,
+      subscribers: subscribers
+    } = state
+
+    # Add the node, notify the node, notify the services.
+    nodes = Map.put(nodes, node, {id, Process.monitor(pid), remote_names})
+    persist_node_names(topology, nodes)
+
+    _ = for {_, pid} <- subscribers, do: send(pid, {:nodeup, node})
+    _ = for {_, name} <- monitors, do: local_monitor_up(local_names, node, id, name)
+    %{state | nodes: nodes}
+  end
+
+  defp delete_node_and_notify(state, node) do
+    %{
+      topology: topology,
+      nodes: nodes,
+      local_names: local_names,
+      subscribers: subscribers
+    } = state
+
+    # Notify the services, remove the node, notify the node.
+    {id, _ref, remote_names} = Map.fetch!(nodes, node)
+    _ = for {name, _} <- remote_names, do: local_monitor_down(local_names, node, id, name)
+
     nodes = Map.delete(nodes, node)
-    update_topology(topology, nodes)
-    Enum.each(subscribers, fn {_ref, pid} -> send(pid, {:nodedown, node}) end)
+    persist_node_names(topology, nodes)
+
+    _ = for {_, pid} <- subscribers, do: send(pid, {:nodedown, node})
     %{state | nodes: nodes}
   end
 
-  defp update_topology(topology, nodes) do
+  defp local_monitor_up(local_names, node, id, name) do
+    local_send(local_names, name, {:named_up, node, id, name})
+  end
+
+  defp local_monitor_down(local_names, node, id, name) do
+    local_send(local_names, name, {:named_down, node, id, name})
+  end
+
+  # Sends a message to the process named `name` in the topology.
+  # Note that we use the pid instead of sending a message to name
+  # to avoid races in case the process dies and we end-up accidentally
+  # messaging the new process.
+  defp local_send(local_names, name, message) do
+    case local_names do
+      %{^name => {pid, _}} -> send(pid, message)
+      %{} -> :error
+    end
+
+    :ok
+  end
+
+  defp maybe_remove_dead_monitor(%{local_names: local_names} = state, name) do
+    case local_names do
+      %{^name => {pid, ref}} ->
+        if Process.alive?(pid) do
```

Comment:
I wonder if `Process.demonitor(ref, [:flush, :info])` would work better here.

Reply:
But if it is alive we don't want to flush it, do we?